src/test/java/com/google/firebase/messaging/FirebaseMessagingTest.java

```diff
@@ -0,0 +1,554 @@
+    TestResponseInterceptor interceptor = new TestResponseInterceptor();
+    messaging.setInterceptor(interceptor);
+
+    TopicManagementResponse result = messaging.subscribeToTopicAsync(
+        ImmutableList.of("id1", "id2"), "test-topic").get();
+    assertEquals(1, result.getSuccessCount());
+    assertEquals(1, result.getFailureCount());
+    assertEquals(1, result.getErrors().size());
+    assertEquals(1, result.getErrors().get(0).getIndex());
+    assertEquals("unknown-error", result.getErrors().get(0).getReason());
+
+    assertNotNull(interceptor.getResponse());
+    HttpRequest request = interceptor.getResponse().getRequest();
+    assertEquals("POST", request.getRequestMethod());
+    assertEquals(TEST_IID_SUBSCRIBE_URL, request.getUrl().toString());
+    assertEquals("Bearer test-token", request.getHeaders().getAuthorization());
+    assertEquals("true", request.getHeaders().get("access_token_auth"));
+
+    ByteArrayOutputStream out = new ByteArrayOutputStream();
+    request.getContent().writeTo(out);
+    assertEquals("{\"to\":\"/topics/test-topic\",\"registration_tokens\":[\"id1\",\"id2\"]}",
+        out.toString());
+  }
+
+  @Test
+  public void testSubscribeError() throws Exception {
+    MockLowLevelHttpResponse response = new MockLowLevelHttpResponse();
+    FirebaseMessaging messaging = initMessaging(response);
+    for (int statusCode : HTTP_ERRORS) {
+      response.setStatusCode(statusCode).setContent("{\"error\": \"test error\"}");
+      TestResponseInterceptor interceptor = new TestResponseInterceptor();
+      messaging.setInterceptor(interceptor);
+      try {
+        messaging.subscribeToTopicAsync(ImmutableList.of("id1", "id2"), "test-topic").get();
+        fail("No error thrown for HTTP error");
+      } catch (ExecutionException e) {
+        assertTrue(e.getCause() instanceof FirebaseMessagingException);
+        FirebaseMessagingException error = (FirebaseMessagingException) e.getCause();
+        assertEquals(getTopicManagementErrorCode(statusCode), error.getErrorCode());
+        assertEquals("test error", error.getMessage());
+        assertTrue(error.getCause() instanceof HttpResponseException);
+      }
+
+      assertNotNull(interceptor.getResponse());
+      HttpRequest request = interceptor.getResponse().getRequest();
+      assertEquals("POST", request.getRequestMethod());
+      assertEquals(TEST_IID_SUBSCRIBE_URL, request.getUrl().toString());
+      assertEquals("Bearer test-token", request.getHeaders().getAuthorization());
+    }
+  }
+
+  @Test
+  public void testInvalidUnsubscribe() {
+    FirebaseOptions options = new FirebaseOptions.Builder()
+        .setCredentials(new MockGoogleCredentials("test-token"))
+        .setProjectId("test-project")
+        .build();
+    FirebaseApp.initializeApp(options);
+    FirebaseMessaging messaging = FirebaseMessaging.getInstance();
+    TestResponseInterceptor interceptor = new TestResponseInterceptor();
```

Comment:
Can 314-319 be done in the `@before` method?

Reply:
Moved to a helper method.